Use C++11 member initializer in CTxMemPoolEntry

maflcko opened the pull request:
This removes a bunch of boilerplate, makes the code easier to read.
Also, C++11 member initialization avoids accidental uninitialized
members.

Can be reviewed with the git option "--word-diff-regex=." or with "git
difftool --tool=meld".

laanwj commented:
Code review ACK fa922062c2eb122e6c4972f20b67723332adcb8c

jonatack commented on src/txmempool.h:
```diff
@@ -115,10 +115,10 @@ class CTxMemPoolEntry
-                    int64_t _nTime, unsigned int _entryHeight,
-                    bool spendsCoinbase,
-                    int64_t nSigOpsCost, LockPoints lp);
+    CTxMemPoolEntry(const CTransactionRef& tx, const CAmount& fee,
```
Feel free to ignore but if you retouch, while changing this function declaration and definition, could also do
```suggestion
    CTxMemPoolEntry(const CTransactionRef& tx, CAmount fee,
```

MarcoFalke replied:
Thanks, done

jonatack reviewed (commented):
Approach ACK

hebasto commented:
Concept ACK since #17786 :)

Commit:
Use C++11 member initializer in CTxMemPoolEntry

This removes a bunch of boilerplate, makes the code easier to read.
Also, C++11 member initialization avoids accidental uninitialized
members.

Can be reviewed with the git option "--word-diff-regex=." or with "git
difftool --tool=meld".

jnewbery commented on src/txmempool.cpp:
```diff
@@ -21,23 +21,26 @@
+      nCountWithAncestors{1},
+      nSizeWithAncestors{GetTxSize()},
+      nModFeesWithAncestors{nFee},
+      nSigOpCostWithAncestors{sigOpCost} {}
```
Sorry for painting the bikeshed, but wouldn't standard project style be to put the body of the function on a new line?

```suggestion
      nSigOpCostWithAncestors{sigOpCost}
  {}
```

MarcoFalke replied:
Clang-format accepts the trailing `{}`, so I think it is fine

jnewbery commented on src/txmempool.cpp:
```diff
@@ -21,23 +21,26 @@
+      entryHeight{entry_height},
+      spendsCoinbase{spends_coinbase},
+      sigOpCost{sigops_cost},
+      feeDelta{0},
```
What do you think about initializing this (and perhaps `nCountWithDescendants` and `nCountWithAncestors`) with default initialization rather than in the initializer list?

MarcoFalke replied:
Thanks, done

jnewbery reviewed (commented):
Concept ACK

shaavan reviewed (approved):
Code Review ACK fa08d4cfb1e3447a76093d46edc1c9de5ceee454

This PR simplifies the code, removing unnecessary parts, at the same time, keeping the same functionality. The involved constants are now initialized through default initialization rather than from the function. The Clang Formatting of the PR is correct. So there are no formatting issues either. Thanks, @MarcoFalke, for catching these redundancies.

jnewbery commented:
Code review ACK fa08d4cfb1e3447a76093d46edc1c9de5ceee454

MarcoFalke commented:
Failing CI can be ignored

JeremyRubin commented on src/txmempool.h:
```diff
@@ -98,27 +95,27 @@ class CTxMemPoolEntry
 
     // Analogous statistics for ancestor transactions
-    uint64_t nCountWithAncestors;
+    uint64_t nCountWithAncestors{1};
```
I don't see a reason to not use the initializer list for these fields, seems more consistent style wise.

MarcoFalke replied:
I changed this because several people asked me to, see https://github.com/bitcoin/bitcoin/pull/23054#discussion_r713030035

hebasto reviewed (commented):
Post-merge ACK fa08d4cfb1e3447a76093d46edc1c9de5ceee454.
